Update log-destination to use the enumeration for LoggingDestination.

rolfyone opened the pull request:
Addresses #1641

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Update log-destination to use the enumeration for LoggingDestination.

Addresses #1641

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 1641-log-destination

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton commented on artemis/src/main/java/tech/pegasys/artemis/cli/options/LoggingOptions.java:
```diff
@@ -54,9 +55,10 @@
+          "Whether a logger is added for the console, the log file, or both (Valid values: ${COMPLETION-CANDIDATES})",
       arity = "1")
-  private String logDestination = DEFAULT_LOG_DESTINATION;
+  private LoggingDestination logDestination = LoggingDestination.get(DEFAULT_LOG_DESTINATION);
```
Wouldn't it make more sense to just use:
```suggestion
  private LoggingDestination logDestination = LoggingDestination.DEFAULT_BOTH;
```

rolfyone replied:
just means its not set off the default - i can if you prefer...

ajsutton replied:
The `DEFAULT_LOG_DESTINATION` constant could also be of type `LoggingDestination`.  Although I really don't know why we've extracted all this stuff as constants and will likely do a pass through to inline them all when I get a chance.  It's so much easier to read in besu where the strings are used directly in the `@Option` annotations.

ajsutton commented on artemis/src/test/resources/complete_config.yaml:
```diff
@@ -32,7 +32,7 @@ eth1-endpoint: "http://localhost:8545"
 log-color-enabled: True
 log-include-events-enabled: True
-log-destination: "both"
+log-destination: "DEFAULT_BOTH"
```
This probably should specify something that's not the default so we can tell it actually got picked up.

rolfyone replied:
its tested in other tests, but we can...

rolfyone replied:
this test is actually super fragile, i'm going to avoid changing that complete_config file...

rolfyone replied:
refactored tests a little bit so that we can see all logging options are picked up from configuration file. I think I've addressed the core of this concern.

ajsutton commented on artemis/src/main/java/tech/pegasys/artemis/cli/options/LoggingOptions.java:
```diff
@@ -81,7 +83,7 @@ public boolean isLogIncludeEventsEnabled() {
 
   public String getLogDestination() {
-    return logDestination;
+    return logDestination.name();
```
The code that uses this value expects the `key` not the enum `name()`.  Currently key is internal only so a getter would be needed. I think it happens to work out at the moment but we should actually refer to the same value.  I suspect the `key` value is no longer relevant and we should just use `.valueOf` to look up the matching value.  Or just pass `LoggingDestination` enum through instead of going to `String` and back if the class path will let us do that.

rolfyone replied:
the project dependencies are pretty ugly. i'll have another go.

ajsutton replied:
I wouldn't try adjusting dependencies to make it happen.  Just make sure we consistently use either name or key to go to string and back.

rolfyone committed:
Added new tests for LoggingOptions, splitting them out from the old BeaconNodeCommandTest suite.

 - new abstract class for commandline tests

 - config file test specific to loggingOptions

 - addressed review feedback.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 1641-log-destination

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 1641-log-destination

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 1641-log-destination

ajsutton commented on artemis/src/test/resources/complete_config.yaml:
```diff
@@ -32,7 +32,7 @@ eth1-endpoint: "http://localhost:8545"
 log-color-enabled: True
 log-include-events-enabled: True
-log-destination: "both"
+log-destination: "BOTH"
```
nit: Can probably undo this change - it should be case insensitive but lowercase will definitely be more common (and more likely to break since the enum name is uppercase)

rolfyone replied:
except the tests don't like it...

ajsutton reviewed (approved):
LGTM.

rolfyone committed:
review feedback

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 1641-log-destination

mbaxter commented on artemis/src/test/java/tech/pegasys/artemis/cli/AbstractBeaconNodeCommandTest.java:
```diff
@@ -0,0 +1,60 @@
+import org.mockito.ArgumentCaptor;
+import tech.pegasys.artemis.util.config.ArtemisConfiguration;
+
+public class AbstractBeaconNodeCommandTest {
```
Should this actually be `abstract`? :D 
```suggestion
public abstract class AbstractBeaconNodeCommandTest {
```

rolfyone replied:
yes :) it should!

mbaxter commented on artemis/src/test/java/tech/pegasys/artemis/cli/options/LoggingOptionsTest.java:
```diff
@@ -0,0 +1,69 @@
+import tech.pegasys.artemis.util.config.ArtemisConfiguration;
+import tech.pegasys.artemis.util.config.LoggingDestination;
+
+public class LoggingOptionsTest extends AbstractBeaconNodeCommandTest {
```
Nice call splitting this out 🎉

mbaxter commented on artemis/src/test/java/tech/pegasys/artemis/cli/options/LoggingOptionsTest.java:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  @Test
+  public void logDestination_ShouldAcceptFileAsDestination() {
```
(nit) I think we usually capitalize like:
```suggestion
  public void logDestination_shouldAcceptFileAsDestination() {
```

rolfyone replied:
gah, i knew that and i still stuffed it up

mbaxter commented on util/src/main/java/tech/pegasys/artemis/util/config/LoggingDestination.java:
```diff
@@ -11,7 +11,7 @@
  */
 
-package tech.pegasys.teku.logging;
+package tech.pegasys.artemis.util.config;
```
Ultimately,`util` is the wrong place for logging-specific constants.  But fixing this issue will require a larger refactor.  Will discuss offline when / how to prioritize refactoring.

rolfyone replied:
sounds good!

mbaxter reviewed (approved):
A few nits, but otherwise LGTM! 🚀

rolfyone committed:
review feedback

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 1641-log-destination
